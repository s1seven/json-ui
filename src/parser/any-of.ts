import { isObject, isArray, omit, get, set, isBoolean } from "lodash";
import { deepMerge } from "../utils/deep-merge";
import { JSONSchema7Value, ExtractObjects } from "../utils/helper-types";
import { JSONSchema7 } from "json-schema";
import Ajv from "ajv";

export const getAnyOfVal = (
  item: JSONSchema7Value
): JSONSchema7[] | undefined => {
  if (!isObject(item) || isArray(item) || !isArray(item.anyOf)) return void 0;
  const { anyOf } = item;
  if (anyOf.findIndex(isBoolean) !== -1)
    throw new Error("booleans not supported in anyOf");
  return anyOf as JSONSchema7[];
};

export const anyOf = (item: JSONSchema7, indices: number[]): JSONSchema7 => {
  const anyOfVal = getAnyOfVal(item);
  if (!anyOfVal) return item;
  return deepMerge<ExtractObjects<JSONSchema7Value>>(
    ...indices.map((idx) => anyOfVal[idx] as ExtractObjects<JSONSchema7Value>),
    omit(
      item as ExtractObjects<JSONSchema7Value>,
      "anyOf"
    ) as ExtractObjects<JSONSchema7Value>
  );
};

export interface AnyOfOption {
  title: string;
  isIllogicalAfterToggle: boolean;
  isSelected: boolean;
}

export const anyOfOptions = (
  schema: JSONSchema7,
  path: string,
  selectedIndices: number[]
): undefined | AnyOfOption[] => {
  const item = path ? (get(schema, path) as JSONSchema7Value) : schema;
  const anyOfVal = getAnyOfVal(item);
  if (!anyOfVal) return void 0;
  const clonedSchema = structuredClone(schema);
  const ajv = new Ajv();
  return anyOfVal.map((val, idx) => {
    const isSelected = selectedIndices.includes(idx);
    const indicesAfterToggle = isSelected
      ? selectedIndices.filter((i) => i !== idx)
      : [...selectedIndices, idx];
    const schemaAfterToggle = anyOf(item as JSONSchema7, indicesAfterToggle);
    set(clonedSchema, path, schemaAfterToggle);
    const hasTypeConflict =
      new Set(indicesAfterToggle.map((i) => anyOfVal[i]?.type).filter(Boolean))
        .size > 1;
    const isIllogicalAfterToggle =
      hasTypeConflict || !(ajv.validateSchema(clonedSchema) as boolean);
    return {
      title: val.title ?? `Option ${idx}`,
      isIllogicalAfterToggle,
      isSelected,
    };
  });
};
