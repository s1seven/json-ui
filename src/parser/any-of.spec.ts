import { JSONSchema7 } from "json-schema";
import { anyOf, anyOfOptions } from "./any-of";

test("anyOf applied correctly", async () => {
  const schema = (await import("../../fixtures/schema-with-any-of.json"))
    .default as unknown as JSONSchema7;
  const result = anyOf(schema, [0, 1]);
  expect(result).toMatchSnapshot();
});

test("anyOf options parsed correctly", async () => {
  const schema = (await import("../../fixtures/schema-with-any-of.json"))
    .default as unknown as JSONSchema7;
  const result = anyOfOptions(schema, "", []);
  expect(result).toMatchSnapshot();
});

test("selected anyOf options parsed correctly", async () => {
  const schema = (await import("../../fixtures/schema-with-any-of.json"))
    .default as unknown as JSONSchema7;
  const selectedIndices = [0, 1];
  const result = anyOfOptions(schema, "", selectedIndices);
  expect(result).toMatchSnapshot();
});
