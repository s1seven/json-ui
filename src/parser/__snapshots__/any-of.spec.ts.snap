// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anyOf applied correctly 1`] = `
{
  "$id": "https://example.com/schemas/product",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "properties": {
    "age": {
      "minimum": 0,
      "type": "number",
    },
    "id": {
      "type": "string",
    },
    "name": {
      "type": "string",
    },
  },
  "required": [
    "name",
    "age",
    "id",
  ],
  "type": "object",
}
`;

exports[`anyOf options parsed correctly 1`] = `
[
  {
    "isIllogicalAfterToggle": false,
    "isSelected": false,
    "title": "Option 0",
  },
  {
    "isIllogicalAfterToggle": false,
    "isSelected": false,
    "title": "Option 1",
  },
  {
    "isIllogicalAfterToggle": false,
    "isSelected": false,
    "title": "Option 2",
  },
  {
    "isIllogicalAfterToggle": false,
    "isSelected": false,
    "title": "Option 3",
  },
]
`;

exports[`selected anyOf options parsed correctly 1`] = `
[
  {
    "isIllogicalAfterToggle": false,
    "isSelected": true,
    "title": "Option 0",
  },
  {
    "isIllogicalAfterToggle": false,
    "isSelected": true,
    "title": "Option 1",
  },
  {
    "isIllogicalAfterToggle": true,
    "isSelected": false,
    "title": "Option 2",
  },
  {
    "isIllogicalAfterToggle": true,
    "isSelected": false,
    "title": "Option 3",
  },
]
`;
